graph/push_v2.go

```diff
@@ -254,8 +252,28 @@ func (p *v2Pusher) pushV2Image(bs distribution.BlobService, img *image.Image) (d
 		Action:   "Pushing",
 	})
 
+	digester := digest.Canonical.New()
+	// HACK: The MultiWriter doesn't write directly to layerUpload because
+	// we must make sure the ReadFrom is used, not Write. Using Write would
+	// send a PATCH request for every Write call.
+	pipeReader, pipeWriter := io.Pipe()
+	compressor := gzip.NewWriter(io.MultiWriter(pipeWriter, digester.Hash()))
+
+	go func() {
+		_, err := io.Copy(compressor, reader)
+		if err == nil {
+			err = compressor.Close()
```

Comment:
The purpose of Close is to flush buffered data and finalize the gzip stream. If we're terminating early due to an error, I don't think it's necessary.